### ISSUE TITLE
Overhead on Thread scheduling/synchronizing?

### ISSUE THREAD

**Issue by soywiz** [jpcsp/jpcsp#424]
I'm not sure on how thread scheduling works. What I have seen is that right now there are a OS thread per PSP thread, (instead of a single thread that was before) but it seems to be a lot of overhead there.

For example Project Diva or Tekken 6 are very slow ingame.

I have managed to double fps in project diva's menu (from 5fps to 10fps) just changing the RuntimeContext.syncIdle : Utilities.sleep() -> to Thread.sleep with the try {} catch outside the loop.
https://github.com/jpcsp/jpcsp/commit/6b413f7de3a7e7fd374af8c0b7fdb4529eae0abb

Not sure how is this implemented but it seems that there is something wrong here. Maybe instead of polling wheter it is idle or not, just wait a signal that wakes the thread? BTW: Why changed you the single threaded model? It improved performance? Or simplified stuff?

**Comment by soywiz** [comment]
In other emulators where I did not have microseconds precission or high sleep overhead, what I did when sleeping was to accumulate microseconds and then do millisecond sleeps.

**Comment by hlide** [comment]
/Off-topic: start
it seems you used TAB instead of SPACE in indentation, it makes the source less readable here (outside editor with the right SPACE number per TAB). May I suggest for your editor to insert SPACEs instead of a TAB?
\Off-topic: end

I didn't know JPCSP uses a thread per PSP thread. I used in my emulators something which is called Fibers (like coroutines) or UMS (64-bit version) and they worked pretty well (PSP registers were in their own fibers). I understand that using real threads will lead to some overheads as PSP uses cooperative theads model and not preemptive threads model because you need to have an "überscheduler" to assure only one thread is running to respect the cooperative nature of PSP threads.

**Comment by soywiz** [comment]
I'm using tabs usually. If you are using tabs just at the begining and consistently it displays just fine everywhere. But since it seems that the project uses spaces, I enforced it to any .editorconfig editors including intelliJ, so I will commit with the correct style.

Fibers are exactly preemtive lightweight threads you can indeed have thousand of thems with less resources. Though they have their own stack so they are not free in terms of memory. But in this case I think it is exactly what is needed. Since PSP have just one core and can't have real multitasking, you won't benefit from having several threads for emulating it. But fibers/threads could allow to stop/resume work at any specific point.

Not sure if java supports fibers out of the box. As far as I know quasar provides fibers but don't know the details.

**Comment by hlide** [comment]
"Like threads, fibers share address space. However, fibers use co-operative multitasking while threads use pre-emptive multitasking. Threads often depend on the kernel's thread scheduler to preempt a busy thread and resume another thread; fibers yield themselves to run another fiber while executing." source: https://en.wikipedia.org/wiki/Fiber_(computer_science).

UMS is even more superior to Fiber but it is only offered in Windows 64-bit (UMS thread blocked by a system call signals the UMS Scheduler so the latter can take a decision like switching to another UMS thread). A Fiber blocked by a system call sits until the system call task is done.

I doubt Java has fibers - the main reason why quasar provides them I think.

**Comment by soywiz** [comment]
<- I said the opposite with the preemtiveness. I do not use that vocabulary so much lately :) But I meant that are the opposite that normal Threads since they cannot execute at once and you can manually schedule them. I used fibers to implement a DIV game studio like engine. But nowadays I'm using bidirectional generators / corroutines implemented at language level in C#/JS/Kotlin that doesn't require OS facilities at all.

For the record:

![s1](https://cloud.githubusercontent.com/assets/570848/20248846/c3a936ec-a9ec-11e6-8643-ac9f0ca8dfb3.png)

![s2](https://cloud.githubusercontent.com/assets/570848/20248845/c38f8440-a9ec-11e6-98fc-acb2f6e7c17f.png)

![s3](https://cloud.githubusercontent.com/assets/570848/20248873/6e3fb964-a9ed-11e6-8498-b99eab4d86ae.png)

![s4](https://cloud.githubusercontent.com/assets/570848/20248872/6e3f97f4-a9ed-11e6-9b1a-c6d33c3c2e8f.png)

![s5](https://cloud.githubusercontent.com/assets/570848/20248871/6e3e5bdc-a9ed-11e6-8c33-d21386b6d239.png)

**Comment by soywiz** [comment]
What do you interpret from that? I have a pretty highend computer running it a i7@3.6 quad core + GTX 970.
So it seems that most threads are waiting for semaphores, vsyncs and sleeps but the fps are pretty low. Also it seemed to have some problems "asking too much" "I'm still sleeping?  I'm still sleeping? Am I awake already?".
But also seems that the GUI thread is doing a lot of work. Drawing should be pretty fast for the computer I have, so I can asume that the displaylist evaluation is the slow part? I did some optimizations in there for jspspemu: http://blog.jspspemu.com/2015/05/10/next-version-huge-graphics-speedup.html
But not sure if that's the problem.

**Comment by hlide** [comment]
glGetTexImage: as I recall, it has slugglish performance whatever the GPU card.
SwapBuffers: I wonder if vsync wait is active.

**Comment by soywiz** [comment]
Well, but if that's for writting down rendered scene to psp's memory so it can be read, that should happen just once per frame. And 512x272 or 480x272 pixels shouldn't be too much. Furthermore I get consistent 60 fps in other parts or games. So that's why I don't really know what's wrong.

**Comment by hlide** [comment]
Using something like that would help a lot: 


minifire interpreter-like 0 7ms vsync on:
![minifire interpreter-like 0 7ms vsync on](https://cloud.githubusercontent.com/assets/56785/20249141/649ce3fe-a9f3-11e6-86bc-7d0f5254b72e.png)
here sceDisplayWaitVblank is more and less waiting on SwapBuffer with active vsync wait. Suppose it isn't the case, you could have an asynchonous or too long drawing conflicting with a SwapBuffer with active vsync wait. 

minifire interpreter-like 0 7ms vsync off:
![minifire interpreter-like 0 7ms vsync off](https://cloud.githubusercontent.com/assets/56785/20249142/69d3cea0-a9f3-11e6-9250-11f8aeb45bc8.png)
here sceDisplayWaitVblank is more and less waiting on SwapBuffer with no active vsync wait

**Comment by soywiz** [comment]
Where can I download that NVIDIA performance analyzer so I can check?

**Comment by hlide** [comment]
It isn't for Jpcsp. that NVIDIA performance analyzer is for C++ projects. Unless you plan to add JNI to use it, this one is useless. As for the link... I guess it may be this one : http://www.geeks3d.com/tag/nvpa/

**Comment by hlide** [comment]
I said: Using SOMETHING LIKE THAT, not THAT. I dunno if there are the same tools under Java.

**Comment by soywiz** [comment]
Oh I see. Not sure. But definitely wouldn't use JNI Directly but JNA that doesn't require to compile C at all nor adding nasty dlls.

BTW I found that "emu.enablegetexture=1" improves it a bit (and it is not marked by default and I think it should), though it is still a bit slow/laggy (went from 10fps to 15fps)

**Comment by hlide** [comment]
Also, did you check that you are not running with internal/external software render?

**Comment by soywiz** [comment]
I'm using OpenGL renderer.
Playing a bit I have noticed that it lags when text is shown and goes 30fps stable ingame when no text is shown but downs to 15~20fps when text is displayed (a single phrase, maybe 20 characters). So maybe there is something related to the texture cache or that each character sis being drawed in different batches recalculating the texture hash or something.

Exporting 3d scene seems to just be dumping text, so maybe it is being drawn in a different list or something related to that?

![image09e52580_09e52500](https://cloud.githubusercontent.com/assets/570848/20249583/99069366-a9fc-11e6-8337-52a7e0c37936.png)

```
newmtl material1
illum 1
Ka 1.000000 1.000000 1.000000
Kd 0.000000 0.000000 0.000000
Ks 0.000000 0.000000 0.000000
map_Kd export/Export-1\Image09E52580_09E52500.png
```

```
mtllib export.mtl
# modelCount=1, vertexCount=1, textureCount=1, normalCount=0
v 269.000000 254.000000 0.000000
vt 0.164063 0.023438
v 269.000000 266.000000 0.000000
vt 0.164063 0.046875
v 275.000000 254.000000 0.000000
vt 0.175781 0.023438
v 275.000000 254.000000 0.000000
vt 0.175781 0.023438
v 269.000000 266.000000 0.000000
vt 0.164063 0.046875
v 275.000000 266.000000 0.000000
vt 0.175781 0.046875
v 275.000000 254.000000 0.000000
vt 0.117188 0.023438
v 275.000000 266.000000 0.000000
vt 0.117188 0.046875
v 281.000000 254.000000 0.000000
vt 0.128906 0.023438
v 281.000000 254.000000 0.000000
vt 0.128906 0.023438
v 275.000000 266.000000 0.000000
vt 0.117188 0.046875
v 281.000000 266.000000 0.000000
vt 0.128906 0.046875
v 281.000000 254.000000 0.000000
vt 0.000000 0.023438
v 281.000000 266.000000 0.000000
vt 0.000000 0.046875
v 287.000000 254.000000 0.000000
vt 0.011719 0.023438
v 287.000000 254.000000 0.000000
vt 0.011719 0.023438
v 281.000000 266.000000 0.000000
vt 0.000000 0.046875
v 287.000000 266.000000 0.000000
vt 0.011719 0.046875
v 287.000000 254.000000 0.000000
vt 0.281250 0.023438
v 287.000000 266.000000 0.000000
vt 0.281250 0.046875
v 293.000000 254.000000 0.000000
vt 0.292969 0.023438
v 293.000000 254.000000 0.000000
vt 0.292969 0.023438
v 287.000000 266.000000 0.000000
vt 0.281250 0.046875
v 293.000000 266.000000 0.000000
vt 0.292969 0.046875
v 293.000000 254.000000 0.000000
vt 0.351563 0.000000
v 293.000000 266.000000 0.000000
vt 0.351563 0.023438
v 299.000000 254.000000 0.000000
vt 0.363281 0.000000
v 299.000000 254.000000 0.000000
vt 0.363281 0.000000
v 293.000000 266.000000 0.000000
vt 0.351563 0.023438
v 299.000000 266.000000 0.000000
vt 0.363281 0.023438
g model1
usemtl material1

f 2/2 1/1 3/3
f 5/5 4/4 6/6
f 8/8 7/7 9/9
f 11/11 10/10 12/12
f 14/14 13/13 15/15
f 17/17 16/16 18/18
f 20/20 19/19 21/21
f 23/23 22/22 24/24
f 26/26 25/25 27/27
f 29/29 28/28 30/30
```

**Comment by soywiz** [comment]
@gid15 Do you know what's happening here?
From what I have investigated I think I found the problem. It seems that each character is displayed in a separated list. And that for each list, the whole window is repainted. Also each list adds 1~2ms of delay. So if it paints the whole 3d scene that takes something like 8ms and then prints 80 characters, that is 80 lists that add 1~2ms each one. So that's why it is so slow. Fixing this drawing the whole scene + characters at once would make it to run smooth as hell.

I have added some traces (that i include)

AbstractDisplayThread calls waitForDisplay and it seems to be completed for each list in this case, repaining for each character after 1 millisecond.
AWTGLCanvas_sceDisplay.paintGL is called and
VideoEngine.update() is called that evaluates the displaylist


```
++++++++++++
VideoEngine.update(listCount=2) started at 1479082321782 since last update 8:
PRIM: 2, 6
PRIM: 49, 4
PRIM: 417, 4
PRIM: 220, 4
PRIM: 352, 4
PRIM: 566, 4
PRIM: 358, 4
PRIM: 861, 4
PRIM: 847, 4
PRIM: 145, 4
PRIM: 155, 4
PRIM: 2238, 3
PRIM: 28, 4
PRIM: 36, 3
PRIM: 420, 3
PRIM: 1019, 4
PRIM: 1020, 3
PRIM: 1020, 3
PRIM: 442, 4
PRIM: 15, 4
PRIM: 6, 4
PRIM: 15, 4
PRIM: 436, 4
PRIM: 15, 4
PRIM: 15, 4
PRIM: 6, 4
PRIM: 6, 4
PRIM: 6, 4
PRIM: 26, 4
PRIM: 40, 4
PRIM: 6, 4
PRIM: 6, 4
PRIM: 6, 4
PRIM: 15, 4
PRIM: 40, 4
PRIM: 26, 4
PRIM: 30, 4
PRIM: 203, 4
PRIM: 34, 4
PRIM: 15, 4
PRIM: 25, 4
PRIM: 25, 4
PRIM: 173, 4
PRIM: 40, 4
PRIM: 15, 4
stallCount: 0, commandCount: 2015
PRIM: 22, 4
PRIM: 36, 4
PRIM: 436, 4
PRIM: 6, 4
PRIM: 10, 4
PRIM: 102, 4
PRIM: 424, 4
PRIM: 15, 4
PRIM: 15, 4
PRIM: 15, 4
PRIM: 15, 4
PRIM: 349, 4
PRIM: 217, 4
PRIM: 217, 4
PRIM: 295, 4
PRIM: 6, 3
stallCount: 0, commandCount: 481
------------
++++++++++++
VideoEngine.update(listCount=1) started at 1479082321785 since last update 3:
PRIM: 6, 3
stallCount: 0, commandCount: 7
------------
++++++++++++
VideoEngine.update(listCount=1) started at 1479082321786 since last update 1:
PRIM: 6, 3
stallCount: 0, commandCount: 7
------------
++++++++++++
VideoEngine.update(listCount=1) started at 1479082321788 since last update 2:
PRIM: 6, 3
stallCount: 0, commandCount: 19
------------
++++++++++++
VideoEngine.update(listCount=1) started at 1479082321789 since last update 1:
PRIM: 6, 3
stallCount: 0, commandCount: 19
------------
++++++++++++
VideoEngine.update(listCount=1) started at 1479082321790 since last update 1:
PRIM: 6, 3
stallCount: 0, commandCount: 7
------------
++++++++++++
VideoEngine.update(listCount=1) started at 1479082321791 since last update 1:
PRIM: 6, 3
stallCount: 0, commandCount: 19
------------
++++++++++++
VideoEngine.update(listCount=1) started at 1479082321793 since last update 2:
PRIM: 6, 3
stallCount: 0, commandCount: 19
------------
++++++++++++
VideoEngine.update(listCount=1) started at 1479082321794 since last update 1:
PRIM: 6, 3
stallCount: 0, commandCount: 19
------------
++++++++++++
VideoEngine.update(listCount=1) started at 1479082321795 since last update 1:
PRIM: 6, 3
stallCount: 0, commandCount: 7
------------
++++++++++++
VideoEngine.update(listCount=1) started at 1479082321797 since last update 2:
PRIM: 6, 3
stallCount: 0, commandCount: 19
------------
++++++++++++
VideoEngine.update(listCount=1) started at 1479082321798 since last update 1:
PRIM: 6, 3
stallCount: 0, commandCount: 19
------------
++++++++++++
VideoEngine.update(listCount=1) started at 1479082321799 since last update 1:
PRIM: 6, 3
stallCount: 0, commandCount: 7
------------
++++++++++++
VideoEngine.update(listCount=1) started at 1479082321800 since last update 1:
PRIM: 6, 3
stallCount: 0, commandCount: 7
------------
++++++++++++
VideoEngine.update(listCount=1) started at 1479082321801 since last update 1:
PRIM: 6, 3
stallCount: 0, commandCount: 7
------------
++++++++++++
VideoEngine.update(listCount=1) started at 1479082321802 since last update 1:
PRIM: 6, 3
stallCount: 0, commandCount: 7
------------
++++++++++++
VideoEngine.update(listCount=1) started at 1479082321804 since last update 2:
PRIM: 6, 3
stallCount: 0, commandCount: 19
------------
...
And so on for each single character on screen on project diva.
```

BTW are you buffering degenerated triangle strips for PRIM groups using the same state? Like this: http://blog.jspspemu.com/2015/05/10/next-version-huge-graphics-speedup.html#nd---degenerate-trianglestrip-buffering I got a huge boost when implemented that.

**Comment by soywiz** [comment]
I think it will require a major refactoring. Since you are mixing list evaluation with rendering and with repainting.
Because probably somewhere game is waiting for a list to be processed. But that process means repainting the whole window + evaluating + rendering.

I think that the correct way is to have:
Game code: generates lists
List evaluator thread: consume lists and generate batches objects (state + vertices) + syncs with game code without actually rendering at all
Rendering thread: this consume batches from list evaluator thread, this actually paints those batches using either software, opengl, webgl, dx, vulkan or whatever.

**Comment by gid15** [comment]
Do the game wait for completion of each GE list? If multiple GE lists are being submitted by an application, they should be rendered in one display. Unless the application is submitting one list and waiting for its completion before submitting the next list.

Multiple PRIM commands are merged when possible (see VideoEngine.checkMultiDraw() ).

**Comment by soywiz** [comment]
Tonight I added some more traces to measure times. Calling AWTGLCanvas_sceDisplay.update, even when performing a very small amount of work is costly when called 80 times per frame (attached traces). It should be called just once.

This seems to ultimately trigger the display. So when immediately is passed or geDirty is passed even when the frame is not done, it triggers the rendering. I think the rendering should happen just once per frame or when the framebuffer has changed. But the framebuffer address is changed inside the lists right? That's why I think it is better to separate the displaylist evaluation from rendering.
```
    public void step(boolean immediately) {
        long now = System.currentTimeMillis();
        if (immediately || now - lastUpdate > 1000 / 60 || geDirty) {
            if (!isOnlyGEGraphics() || VideoEngine.getInstance().hasDrawLists()) {
                if (geDirty || detailsDirty || displayDirty) {
                    detailsDirty = false;
                    displayDirty = false;
                    geDirty = false;

                    asyncDisplayThread.display();
                }
            }
            lastUpdate = now;
        }
    }
```

```
AWTGLCanvas_sceDisplay.update+paint()
AWTGLCanvas_sceDisplay.paintGL()
++++++++++++
VideoEngine.update(listCount=2) started at 1479113635730 since last update 8:
PRIM: 2, 6
PRIM: 49, 4
...
PRIM: 25, 4
PRIM: 173, 4
PRIM: 40, 4
stallCount: 0, commandCount: 2014
PRIM: 15, 4
PRIM: 36, 4
...
PRIM: 217, 4
PRIM: 217, 4
PRIM: 295, 4
PRIM: 6, 3
stallCount: 0, commandCount: 460
------------
sceDisplay.canvas.swapBuffers()
AWTGLCanvas_sceDisplay.update+paint TOOK: 1.716339ms
AWTGLCanvas_sceDisplay.update+paint()
AWTGLCanvas_sceDisplay.paintGL()
++++++++++++
VideoEngine.update(listCount=1) started at 1479113635732 since last update 2:
PRIM: 6, 3
stallCount: 0, commandCount: 7
------------
sceDisplay.canvas.swapBuffers()
AWTGLCanvas_sceDisplay.update+paint TOOK: 0.550116ms
AWTGLCanvas_sceDisplay.update+paint()
AWTGLCanvas_sceDisplay.paintGL()
++++++++++++
VideoEngine.update(listCount=1) started at 1479113635733 since last update 1:
PRIM: 6, 3
stallCount: 0, commandCount: 7
------------
sceDisplay.canvas.swapBuffers()
AWTGLCanvas_sceDisplay.update+paint TOOK: 0.441742ms
AWTGLCanvas_sceDisplay.update+paint()
AWTGLCanvas_sceDisplay.paintGL()
++++++++++++
VideoEngine.update(listCount=1) started at 1479113635734 since last update 1:
PRIM: 6, 3
stallCount: 0, commandCount: 19
------------
sceDisplay.canvas.swapBuffers()
AWTGLCanvas_sceDisplay.update+paint TOOK: 0.521102ms
AWTGLCanvas_sceDisplay.update+paint()
AWTGLCanvas_sceDisplay.paintGL()
++++++++++++
VideoEngine.update(listCount=1) started at 1479113635736 since last update 2:
PRIM: 6, 3
stallCount: 0, commandCount: 19
------------
sceDisplay.canvas.swapBuffers()
AWTGLCanvas_sceDisplay.update+paint TOOK: 0.393103ms
AWTGLCanvas_sceDisplay.update+paint()
AWTGLCanvas_sceDisplay.paintGL()
++++++++++++
VideoEngine.update(listCount=1) started at 1479113635737 since last update 1:
PRIM: 6, 3
stallCount: 0, commandCount: 7
------------
sceDisplay.canvas.swapBuffers()
AWTGLCanvas_sceDisplay.update+paint TOOK: 0.376036ms
AWTGLCanvas_sceDisplay.update+paint()
AWTGLCanvas_sceDisplay.paintGL()
++++++++++++
VideoEngine.update(listCount=1) started at 1479113635738 since last update 1:
PRIM: 6, 3
stallCount: 0, commandCount: 19
------------
sceDisplay.canvas.swapBuffers()
AWTGLCanvas_sceDisplay.update+paint TOOK: 0.388551ms
AWTGLCanvas_sceDisplay.update+paint()
AWTGLCanvas_sceDisplay.paintGL()
++++++++++++
VideoEngine.update(listCount=1) started at 1479113635739 since last update 1:
PRIM: 6, 3
stallCount: 0, commandCount: 19
------------
sceDisplay.canvas.swapBuffers()
AWTGLCanvas_sceDisplay.update+paint TOOK: 0.367502ms
AWTGLCanvas_sceDisplay.update+paint()
AWTGLCanvas_sceDisplay.paintGL()
++++++++++++
VideoEngine.update(listCount=1) started at 1479113635740 since last update 1:
PRIM: 6, 3
stallCount: 0, commandCount: 19
------------
sceDisplay.canvas.swapBuffers()
AWTGLCanvas_sceDisplay.update+paint TOOK: 0.43776ms
AWTGLCanvas_sceDisplay.update+paint()
AWTGLCanvas_sceDisplay.paintGL()
++++++++++++
VideoEngine.update(listCount=1) started at 1479113635741 since last update 1:
PRIM: 6, 3
stallCount: 0, commandCount: 7
------------
sceDisplay.canvas.swapBuffers()
AWTGLCanvas_sceDisplay.update+paint TOOK: 0.468196ms
AWTGLCanvas_sceDisplay.update+paint()
AWTGLCanvas_sceDisplay.paintGL()
++++++++++++
VideoEngine.update(listCount=1) started at 1479113635743 since last update 2:
PRIM: 6, 3
stallCount: 0, commandCount: 19
------------
sceDisplay.canvas.swapBuffers()
AWTGLCanvas_sceDisplay.update+paint TOOK: 0.47104ms
AWTGLCanvas_sceDisplay.update+paint()
AWTGLCanvas_sceDisplay.paintGL()
++++++++++++
VideoEngine.update(listCount=1) started at 1479113635744 since last update 1:
PRIM: 6, 3
stallCount: 0, commandCount: 19
------------
sceDisplay.canvas.swapBuffers()
AWTGLCanvas_sceDisplay.update+paint TOOK: 0.462791ms
AWTGLCanvas_sceDisplay.update+paint()
AWTGLCanvas_sceDisplay.paintGL()
++++++++++++
VideoEngine.update(listCount=1) started at 1479113635745 since last update 1:
PRIM: 6, 3
stallCount: 0, commandCount: 7
------------
sceDisplay.canvas.swapBuffers()
AWTGLCanvas_sceDisplay.update+paint TOOK: 0.403627ms
AWTGLCanvas_sceDisplay.update+paint()
AWTGLCanvas_sceDisplay.paintGL()
++++++++++++
VideoEngine.update(listCount=1) started at 1479113635746 since last update 1:
PRIM: 6, 3
stallCount: 0, commandCount: 7
------------
sceDisplay.canvas.swapBuffers()
AWTGLCanvas_sceDisplay.update+paint TOOK: 0.461369ms
AWTGLCanvas_sceDisplay.update+paint()
AWTGLCanvas_sceDisplay.paintGL()
++++++++++++
VideoEngine.update(listCount=1) started at 1479113635747 since last update 1:
PRIM: 6, 3
stallCount: 0, commandCount: 7
------------
sceDisplay.canvas.swapBuffers()
AWTGLCanvas_sceDisplay.update+paint TOOK: 0.369494ms
AWTGLCanvas_sceDisplay.update+paint()
AWTGLCanvas_sceDisplay.paintGL()
++++++++++++
VideoEngine.update(listCount=1) started at 1479113635749 since last update 2:
PRIM: 6, 3
stallCount: 0, commandCount: 7
------------
sceDisplay.canvas.swapBuffers()
AWTGLCanvas_sceDisplay.update+paint TOOK: 0.452552ms
AWTGLCanvas_sceDisplay.update+paint()
AWTGLCanvas_sceDisplay.paintGL()
++++++++++++
VideoEngine.update(listCount=1) started at 1479113635750 since last update 1:
PRIM: 6, 3
stallCount: 0, commandCount: 19
------------
sceDisplay.canvas.swapBuffers()
AWTGLCanvas_sceDisplay.update+paint TOOK: 0.403627ms
AWTGLCanvas_sceDisplay.update+paint()
AWTGLCanvas_sceDisplay.paintGL()
++++++++++++
VideoEngine.update(listCount=1) started at 1479113635751 since last update 1:
PRIM: 6, 3
stallCount: 0, commandCount: 7
------------
sceDisplay.canvas.swapBuffers()
AWTGLCanvas_sceDisplay.update+paint TOOK: 0.367787ms
AWTGLCanvas_sceDisplay.update+paint()
AWTGLCanvas_sceDisplay.paintGL()
++++++++++++
VideoEngine.update(listCount=1) started at 1479113635752 since last update 1:
PRIM: 6, 3
stallCount: 0, commandCount: 7
------------
sceDisplay.canvas.swapBuffers()
AWTGLCanvas_sceDisplay.update+paint TOOK: 0.396516ms
AWTGLCanvas_sceDisplay.update+paint()
AWTGLCanvas_sceDisplay.paintGL()
++++++++++++
VideoEngine.update(listCount=1) started at 1479113635753 since last update 1:
PRIM: 6, 3
stallCount: 0, commandCount: 7
------------
sceDisplay.canvas.swapBuffers()
AWTGLCanvas_sceDisplay.update+paint TOOK: 0.345316ms
AWTGLCanvas_sceDisplay.update+paint()
AWTGLCanvas_sceDisplay.paintGL()
++++++++++++
VideoEngine.update(listCount=1) started at 1479113635754 since last update 1:
PRIM: 6, 3
PRIM: 24, 3
stallCount: 0, commandCount: 63
------------
sceDisplay.canvas.swapBuffers()
AWTGLCanvas_sceDisplay.update+paint TOOK: 0.833707ms
AWTGLCanvas_sceDisplay.update+paint()
AWTGLCanvas_sceDisplay.paintGL()
++++++++++++
VideoEngine.update(listCount=1) started at 1479113635755 since last update 1:
PRIM: 6, 3
stallCount: 0, commandCount: 7
------------
sceDisplay.canvas.swapBuffers()
AWTGLCanvas_sceDisplay.update+paint TOOK: 0.416712ms
AWTGLCanvas_sceDisplay.update+paint()
AWTGLCanvas_sceDisplay.paintGL()
++++++++++++
VideoEngine.update(listCount=1) started at 1479113635756 since last update 1:
PRIM: 36, 3
stallCount: 0, commandCount: 7
------------
sceDisplay.canvas.swapBuffers()
AWTGLCanvas_sceDisplay.update+paint TOOK: 0.484124ms
AWTGLCanvas_sceDisplay.update+paint()
AWTGLCanvas_sceDisplay.paintGL()
++++++++++++
VideoEngine.update(listCount=1) started at 1479113635758 since last update 2:
PRIM: 30, 3
stallCount: 0, commandCount: 7
------------
sceDisplay.canvas.swapBuffers()
AWTGLCanvas_sceDisplay.update+paint TOOK: 0.630045ms
AWTGLCanvas_sceDisplay.update+paint()
AWTGLCanvas_sceDisplay.paintGL()
++++++++++++
VideoEngine.update(listCount=1) started at 1479113635759 since last update 1:
PRIM: 18, 3
stallCount: 0, commandCount: 47
------------
sceDisplay.canvas.swapBuffers()
AWTGLCanvas_sceDisplay.update+paint TOOK: 0.443734ms
AWTGLCanvas_sceDisplay.update+paint()
AWTGLCanvas_sceDisplay.paintGL()
++++++++++++
VideoEngine.update(listCount=1) started at 1479113635760 since last update 1:
PRIM: 18, 3
stallCount: 0, commandCount: 7
------------
sceDisplay.canvas.swapBuffers()
AWTGLCanvas_sceDisplay.update+paint TOOK: 0.504605ms
AWTGLCanvas_sceDisplay.update+paint()
AWTGLCanvas_sceDisplay.paintGL()
++++++++++++
VideoEngine.update(listCount=1) started at 1479113635761 since last update 1:
PRIM: 36, 3
stallCount: 0, commandCount: 7
------------
sceDisplay.canvas.swapBuffers()
AWTGLCanvas_sceDisplay.update+paint TOOK: 0.471894ms
AWTGLCanvas_sceDisplay.update+paint()
AWTGLCanvas_sceDisplay.paintGL()
++++++++++++
VideoEngine.update(listCount=1) started at 1479113635762 since last update 1:
PRIM: 42, 3
stallCount: 0, commandCount: 47
------------
sceDisplay.canvas.swapBuffers()
AWTGLCanvas_sceDisplay.update+paint TOOK: 0.725902ms
AWTGLCanvas_sceDisplay.update+paint()
AWTGLCanvas_sceDisplay.paintGL()
++++++++++++
VideoEngine.update(listCount=1) started at 1479113635763 since last update 1:
PRIM: 42, 3
stallCount: 0, commandCount: 7
------------
sceDisplay.canvas.swapBuffers()
AWTGLCanvas_sceDisplay.update+paint TOOK: 0.567467ms
AWTGLCanvas_sceDisplay.update+paint()
AWTGLCanvas_sceDisplay.paintGL()
++++++++++++
VideoEngine.update(listCount=1) started at 1479113635764 since last update 1:
PRIM: 12, 3
stallCount: 0, commandCount: 7
------------
sceDisplay.canvas.swapBuffers()
AWTGLCanvas_sceDisplay.update+paint TOOK: 0.421831ms
AWTGLCanvas_sceDisplay.update+paint()
AWTGLCanvas_sceDisplay.paintGL()
++++++++++++
VideoEngine.update(listCount=1) started at 1479113635765 since last update 1:
PRIM: 18, 3
stallCount: 0, commandCount: 7
------------
sceDisplay.canvas.swapBuffers()
AWTGLCanvas_sceDisplay.update+paint TOOK: 0.607573ms
AWTGLCanvas_sceDisplay.update+paint()
AWTGLCanvas_sceDisplay.paintGL()
++++++++++++
VideoEngine.update(listCount=1) started at 1479113635766 since last update 1:
PRIM: 24, 3
stallCount: 0, commandCount: 47
------------
sceDisplay.canvas.swapBuffers()
AWTGLCanvas_sceDisplay.update+paint TOOK: 0.461085ms
...
```

**Comment by gid15** [comment]
Could you upload a log file after enabling the log level to DEBUG for a few seconds when the rendering is slow (Menu Debug->Tools->Logger->Show Logger / change from INFO to DEBUG)?
I don't think it would be the right approach to first save the content of the GE list and render it some time later:
1. The list including all sublists, all vertices data, all texture data would need to be copied into that intermediate object as the application is free to overwrite the data after the list is completed. The performance impact could be high.
2. After a list is completed, the application assumes the display is rendered and can perform operations on the GE memory: the most common is probably to simply call sceDisplaySetFrameBuf, but I do remember having seen applications doing other kind of memory operations, like copying (using memcpy) the rendered screen somewhere else.

The right approach from my point of view is to do exactly the same as the PSP does: when a list is submitted, start to parse the list and render immediately as instructed by the list. This is the approach used in the external software renderer. Have you tried it and checked his performance on this specific game?

I had the idea to mix both software and OpenGL renderers when rendering a list: for example, render most of the list through OpenGL (better performance) but use the software renderer for the cases that are hard to implement through OpenGL (e.g. some filters applied on the whole framebuffer or complex framebuffer copy or... processing small GE lists like here). This was just an idea up to now and I've not tried it out.

**Comment by soywiz** [comment]
I though that the display lists were decoded by the MediaEngine CPU and emitting commands asynchronously to the GraphicsCore.
I didn't mean to copy all the lists.

In jspspemu I have the following architecture:

List evaluator is pretty simple just manipulating an int[256] and generating batches. In the switch I have just PRIM + MATRICES, the rest just changes values in that int array:
https://github.com/jspspemu/jspspemu/blob/d9c3ebf40a91abfd891becd703f6e832e082f07c/source/src/core/gpu/gpu_core.ts#L112

The state object tree is just a view of that int[256] so you can access stuff by name, but when generating a batch, the state is just a small chunk of memory of 1024 bytes:
https://github.com/jspspemu/jspspemu/blob/d9c3ebf40a91abfd891becd703f6e832e082f07c/source/src/core/gpu/gpu_state.ts#L253

Then I have a DrawBuffer to arrays one for vertices and other for indices:
https://github.com/jspspemu/jspspemu/blob/d9c3ebf40a91abfd891becd703f6e832e082f07c/source/src/core/gpu/gpu_vertex.ts#L211

And a Batch with state + texture + references to the index buffer:
https://github.com/jspspemu/jspspemu/blob/d9c3ebf40a91abfd891becd703f6e832e082f07c/source/src/core/gpu/gpu_vertex.ts#L298

The list evaluator just populates very quickly the DrawBuffer + Batches. And that information is passed to the renderer, that checks for changes in the state and performs the drawing in a single step:
https://github.com/jspspemu/jspspemu/blob/d9c3ebf40a91abfd891becd703f6e832e082f07c/source/src/core/gpu/webgl/webgl_driver.ts

In HTML you have to render everything in a single step before returning to the event loop because flipping occurs automatically.
Also this allows to have the displaylist evaluato in one thread and the renderer in the main thread, so better using cores.

-----

About the performance, I have investigated a bit more:

With this in AWTGLCanvas_sceDisplay:
```
@Override
public void update(Graphics g) {
    long start = System.nanoTime();
    super.update(g);
    long end = System.nanoTime();
    log.warn("AWTGLCanvas_sceDisplay.update:" + (end - start));
}

@Override
protected void paintGL() {
    long start = System.nanoTime();
    _paintGL();
    long end = System.nanoTime();
    log.warn("AWTGLCanvas_sceDisplay.paintGL:" + (end - start));
}
```

It outputs this:
```
47584 [GUI] WARN  hle.sceDisplay - AWTGLCanvas_sceDisplay.update:834277
47586 [GUI] WARN  hle.sceDisplay - AWTGLCanvas_sceDisplay.paintGL:367503
47587 [GUI] WARN  hle.sceDisplay - AWTGLCanvas_sceDisplay.update:946917
47589 [GUI] WARN  hle.sceDisplay - AWTGLCanvas_sceDisplay.paintGL:227555
47589 [GUI] WARN  hle.sceDisplay - AWTGLCanvas_sceDisplay.update:846792
47591 [GUI] WARN  hle.sceDisplay - AWTGLCanvas_sceDisplay.paintGL:226987
47592 [GUI] WARN  hle.sceDisplay - AWTGLCanvas_sceDisplay.update:852766
47594 [GUI] WARN  hle.sceDisplay - AWTGLCanvas_sceDisplay.paintGL:233245
47594 [GUI] WARN  hle.sceDisplay - AWTGLCanvas_sceDisplay.update:924161
```

As you can see the update takes way longer than paintGL, so even optimizing paintGL is not enough.
update calls paint and paint looks like this:

```
public final void paint(Graphics g) {
	LWJGLException exception = null;
	synchronized ( SYNC_LOCK ) {
		if ( !isDisplayable() )
			return;
		try {
			if ( peer_info == null ) {
				this.peer_info = implementation.createPeerInfo(this, pixel_format, attribs);
			}
			peer_info.lockAndGetHandle();
			try {
				if ( context == null ) {
					this.context = new ContextGL(peer_info, attribs, drawable != null ? (ContextGL)((DrawableLWJGL)drawable).getContext() : null);
					first_run = true;
				}

				if ( reentry_count == 0 )
					context.makeCurrent();
				reentry_count++;
				try {
					if ( update_context ) {
						context.update();
						update_context = false;
					}
					if ( first_run ) {
						first_run = false;
						initGL();
					}
					paintGL();
				} finally {
					reentry_count--;
					if ( reentry_count == 0 )
						context.releaseCurrent();
				}
			} finally {
				peer_info.unlock();
			}
		} catch (LWJGLException e) {
			exception = e;
		}
	}
	if ( exception != null )
		exceptionOccurred(exception);
}
```

It makes the current thread the one owning GL static calls. And that's a lot of overhead. Thus calling paint multiple times will add an overhead of something like 0,6ms. Since this game seems to create a single list per character and somehow the display is triggered, it is sometimes called 100 times per frame. Thus 0,6ms + 0,2ms executing it = 0,8ms * 100 = 80ms per frame which is not good.

BTW: I tried to enable DEBUG, but I don't think that the information shown is relevant since the hugest overhead seems to be the GL.makeCurrent or something on the paint function.

Do you know what are PPSSPP guys doing on this matter?

**Comment by hlide** [comment]
Sowiz, I think there is a misconception I read and made me frown out several times in your posts. Media Engine (ME) has nothing to do with GPU. I guess you mean Graphics Engine (GE), right? ME is only used for stuff like Audio and Video Codecs and has nothing with Display and GE.

As for what you're speaking about, do you mean PaintGL is doing a triplet (CAPTURE TARGET, RENDER TARGET, PRESENT TARGET) 100 times per frame? If so, yes I guess CAPTURE and PRESENT TARGET are excessive as i would expect them to be done once per frame. UNLESS you need to capture into memory EACH rendering but that's another story. What you propose sounds like a hack, a trick which may not work accurately for all the games. The situation you seem to meet here looks like a poorly coded game which doesn't seem to exploit GE in its best capacity.

**Comment by soywiz** [comment]
I though that the extra CPU was used in that case too. So the displaylist is evaluated in the main CPU?
The game seems to generate a list, and somehow the display is triggered (not sure if the thread was suspended or what, but that's the case). In PPSSPP the game runs smoothly. I don't think the game is capturing the memory at all, in fact it doesn't seems to call the synchronizations functions for lists. So maybe it just waits for vsync, and then emits all the lists, not sure.

I have tried to hack the lwjgl (disabling the package sealing) and changing the paint method that was final to this:

```
public void paint(Graphics g) {
    try {
        if (this.context == null) {
            if(this.peer_info == null) {
                this.peer_info = implementation.createPeerInfo(this, this.pixel_format, this.attribs);
            }
            this.peer_info.lockAndGetHandle();
            this.context = new ContextGL(this.peer_info, this.attribs, this.drawable != null ? (ContextGL) ((DrawableLWJGL) this.drawable).getContext() : null);
            this.context.makeCurrent();
            this.initGL();
        }
        paintGL();
    } catch (LWJGLException e) {
        e.printStackTrace();
    }
}
```

Since GL is just used from one thread, this works. But even when that the makeCurrent overhead is gone, it hasn't improved it.

In that case I have uploaded the debug log (not sure if was properly configured for your requirements):

https://gist.github.com/soywiz/56a31555068d4211eeb32d72227b90b6

**Comment by hlide** [comment]
It depends upon what side. I thought you were speaking about hardware side (SC, ME, GE). Displaylist is a GE stuff. I don't remember how it is done in JPCSP. If JPCSP is referring a ME thread to make this displaylist evaluation, this is plainly wrong.

**Comment by soywiz** [comment]
I don't think JPCSP is naming it wrong. That was just what I thought all this time, though never checked it. So I could be totally wrong about it.
http://hitmen.c02.at/files/yapspd/psp_doc/chap2.html#sec2
So where is the GE? Or it is just a system thread on CPU Core?
I though the Graphics Core was not a normal cpu but a GPU (doing rasterizing + vertex transformations + pixel blitting).
Since the ge API has a stall pointer, I thought that it was processed in parallel by another generic CPU (and since PSP is dual core (including the Media Engine I thought that ME was that component)).

**Comment by hlide** [comment]
I eat a lot of OpenGL coding and you know what? that's the worst and dumb interface and implementation. It is a bloated software and has a massive syncpoint in the CPU side (state and context inside only ONE thread) which makes a serious bottleneck per draw call - especially when calling one hundred draws per frame. GE may look like an OpenGL but it doesn't work the same - you could say OpenGL and GE are antagonist in the way they work really.

**Comment by hlide** [comment]
GE is a system which does everything about graphics : including executing GE commands in the GE queue. I don't understand why you want a CPU. The details inside the GE system - maybe there a (third) CPU to handle those commands and jumps, etc. but that kind of detail doesn't matter.

**Comment by soywiz** [comment]
I see. Thanks for the explanation. It has passed sometime since I last worked on a PSP Emulator. And never discussed those details, but tried to get stuff working based on RE work and guessing. So I really appreciates it.

Returning to the main issue:
I think there is too much overhead in the thread scheduler + something wrong to do that many updates. I think the slowness is caused by both. Though in this case specially by having that many updates. I will try to investigate more in the following days. But if the log dump provides useful information to determine why, that would be great.

**Comment by hlide** [comment]
I had a pdf which gave some more details inside the GE workings in a bitbucket message with gid15 but I cannot retrieve those messages any longer.

**Comment by hlide** [comment]
For me it was a long time ago so I may not totally be sure how the sceGe API work.  A call to sceGeDrawSync/sceGeListSync can be blocking or not due to its argument. Check if their argument when those functions are called per letter drawing. If block is true, 100 of them will be clearly lead to overhead translated into OpenGL renderings.

**Comment by soywiz** [comment]
That pdf would be cool, if you can find it in the future.

Regarding to that. This game is not calling sceGeDrawSync/sceGeListSync, but just sceDisplayWaitVblankStart. So probably waiting for vblank then emitting all lists. But not waiting for completion at all. At least the a breakpoint in sceDisplayWaitVblankStart stops the code, but not on sceGeDrawSync/sceGeListSync.

**Comment by hlide** [comment]
sceDisplayWaitVblankStart is not GE stuff (you could say there are totally separate matter). Before rendering the next frame, I world call sceGeDrawSync to make sure all the rendering is complete. I'm puzzled.

**Comment by soywiz** [comment]
I have tried to reduce the "update" count to just once per frame, but the game goes even slower (1fps). So I suspect that the game uses the FINISH callback as a feedback mecanism to continue generating lists. And if  reducing the setcontext opengl overhead didn't help, maybe it is regarded to taking too long to execute that callback (maybe waking up a thread or something by the thread scheduler? Do not remember how that worked)

**Comment by hlide** [comment]
Probably. FINISH callback is another way to wait for completion. And yes, it means an OpenGL draw per FINISH callback. And as I told you, OpenGL draw call are hell slow because of the OpenGL state syncing, no wonder having 100 of them is too much... i wonder if Vulkan would be better in this regard (Vulkan is - probably - more compliant in the way GE works).

**Comment by hlide** [comment]
I understand that what you wanted to do is more or less like transforming a set of OpenGL single draws into one OpenGL multidraw. But how do you handle FINISH callback or calls to sceGeDrawSync/sceGeListSync? when and how to decide that you must call the multidraw?

**Comment by soywiz** [comment]
Yeah, but PPSSPP as far as I know is using opengl and this game works perfectly. Also, reducing the overhead while drawing did not improved a single fps, so maybe the interrupt is not executed fast enough, and it takes several milliseconds so that was the bottleneck. I'm still trying to figure out how the thread scheduling + interrupts + callbacks works in jpcsp right now. So the jwgl hack + fixing the scheduler (if that's the problem) would help a lot.

---

Regarding to multidraw. When I reach the FINISH command, I enqueue that immediately without drawing. Since I already evaluated the list, that's ok. It is not rendered yet, but already has the information needed to render it later in a batch to complete the frame. (Unless we are doing offscreen rendering.) which should imply several renders in a single frame.

**Comment by soywiz** [comment]
@gid15 What does '`jpcsp.Allegrex.compiler.RuntimeSyncThread`? It seems to just sleep in small steps right? I commented out the main while and game still works. But not sure why is this there.

**Comment by soywiz** [comment]
It seems that the game is executing the FINISH callback in the `user_main` thread after a syscall. Maybe it takes too much time to reach that syscall? Should it be executed in other thread sleeping and waiting for CB? Is there a way to know which syscall was it without DEBUG traces? Enabling DEBUG just caused to not reach it because of the overhead.

![geinterrupthandler](https://cloud.githubusercontent.com/assets/570848/20326210/148c56ec-ab88-11e6-9082-6337699bff1e.png)

**Comment by gid15** [comment]
The DEBUG log is almost empty, at least it doesn't show any drawing activity or any call to sceDisplay/sceGe... Have you maybe tweaked some entries in your LogSettings.xml and explicitely disabled some entries?

**Comment by soywiz** [comment]
Check this one:
https://gist.github.com/soywiz/25d67e532cca72938660e22b047c612c

I just copied the log window output, but there is a log.txt file, searching for user_main:

```
04:09:34 DEBUG  runtime - user_main - syncEmulator immediately=true
04:09:34 DEBUG hle.sceDisplay - GUI - sceDisplay.paintGL - start display
04:09:34 DEBUG       ge - GUI - executeList PspGeList[id=0x13, status=PSP_GE_LIST_DRAWING, list=0x08F19980, pc=0x08F19980, stall=0x00000000, cbid=0x0, ended=false, finished=false, paused=false, restarted=false, reset=false]
04:09:34 DEBUG       ge - GUI - base 0x08000000
04:09:34 DEBUG  runtime - user_main - syncEmulator immediately=true
04:09:34 DEBUG       ge - GUI - vaddr 08fc3a60
04:09:34 DEBUG       ge - GUI - vtype GU_TEXTURE_32BITF|GU_COLOR_8888|GU_NORMAL_32BITF|GU_VERTEX_32BITF|GU_TRANSFORM_3D size=36
04:09:34 DEBUG hle.sceDisplay - GUI - hleDisplaySetGeBuf topaddr=0x04000000, bufferwidth=512, pixelformat=3, copyGE=false, with=480, height=272
04:09:34 DEBUG       ge - GUI - Reloading GE Memory (0x04000000-0x04088000) to screen (480x272)
04:09:34 DEBUG       ge - GUI - GETexture.copyTextureToScreen GETexture[0x04000000-0x04088000, 480x272 (texture 1024x1024), bufferWidth=512, pixelFormat=3(PSM_8888)] at 0x0
04:09:34 DEBUG       ge - GUI - Texture.hashCode: 131072 bytes
04:09:34 DEBUG  runtime - user_main - syncEmulator immediately=true
04:09:34 DEBUG       ge - GUI - tflush Reusing cached texture 94
04:09:34 DEBUG       ge - GUI - prim 6 triangle (18 vertices)
04:09:34 DEBUG       ge - GUI - sceGuEnable(GU_SCISSOR_TEST)
04:09:34 DEBUG       ge - GUI - sceGuDisable(RE_TEXTURE_GEN_S)
04:09:34 DEBUG       ge - GUI - sceGuDisable(RE_TEXTURE_GEN_T)
04:09:34 DEBUG       ge - GUI - sceGuDisable(RE_COLOR_MATERIAL)
04:09:34 DEBUG       ge - GUI - Texture Mipmap base=0, max=0, validNumberMipmaps=0
04:09:34 DEBUG       ge - GUI - Not reading Vertex, all native at 0x08FC3A60
04:09:34 DEBUG  runtime - user_main - syncEmulator immediately=true
04:09:34 DEBUG       ge - GUI - tsync wait for rendering completion.
04:09:34 DEBUG hle.sceAudio - user_main - free buffer 21
04:09:34 DEBUG       ge - GUI - finish (0)
04:09:34 DEBUG hle.sceAudio - user_main - free buffer 27
04:09:34 DEBUG hle.sceAudio - user_main - free buffer 15
04:09:34 DEBUG hle.sceGe_user - GUI - Scheduling Async Callback uid:0x6C, name:'GeCallbackFinish', thread:'user_main', PC:0x089D1F94, $a0:0x00000000, $a1:0x00000000, $a2:0x08E31A50, listId=0x13, listPc=0x08F1999C, behavior=1, signalId=0x0
04:09:34 DEBUG hle.sceAudio - user_main - hleAudioBlockingOutput SoundChannel[6](sourceSampleOffset=510, restLength=3458, buffers queued=14, isOutputBlock=false, Stereo, reserved=true, sampleLength=256, sampleRate=44100, busy=true)
04:09:34 DEBUG       ge - GUI - end pc=0x08F1999C
04:09:34 DEBUG hle.sceAudio - user_main - doAudioOutput(SoundChannel[6](sourceSampleOffset=126, restLength=3458, buffers queued=14, isOutputBlock=false, Stereo, reserved=true, sampleLength=256, sampleRate=44100, busy=true), 0x08C7AC00)
04:09:34 DEBUG       ge - GUI - FINISH / SIGNAL / END reached, waiting for Sync (PspGeList[id=0x13, status=PSP_GE_LIST_DRAWING, list=0x08F19980, pc=0x08F1999C, stall=0x00000000, cbid=0x0, ended=true, finished=true, paused=true, restarted=false, reset=false])
04:09:34 DEBUG hle.sceAudio - user_main - alSourceQueueBuffer buffer=15, SoundChannel[6](sourceSampleOffset=126, restLength=3714, buffers queued=15, isOutputBlock=false, Stereo, reserved=true, sampleLength=256, sampleRate=44100, busy=true)
04:09:34 DEBUG hle.sceGe_user - GUI - hleGeListSyncDone list PspGeList[id=0x13, status=PSP_GE_LIST_DONE, list=0x08F19980, pc=0x08F1999C, stall=0x00000000, cbid=0x0, ended=true, finished=true, paused=true, restarted=false, reset=false], done
04:09:34 DEBUG       ge - GUI - sceGuDisable(RE_TEXTURE_GEN_S)
04:09:34 DEBUG hle.ThreadManForUser - user_main - -------------------- unblock SceUID=78 name:'CRI Wave out' caller:'sceAudio.hleAudioBlockingOutput'
04:09:34 DEBUG       ge - GUI - sceGuDisable(RE_TEXTURE_GEN_T)
04:09:34 DEBUG hle.ThreadManForUser - user_main - Triggering Interrupt GE(0x19) at 0x089D1F94
04:09:34 DEBUG hle.sceDisplay - GUI - sceDisplay.paintGL - saving the GE to memory 0x04000000
04:09:34 DEBUG hle.ThreadManForUser - user_main - Calling InterruptHandler 0x089D1F94(0x00000000,0x08E31A50,0x08F1999C)
04:09:34 DEBUG       ge - GUI - GETexture.copyScreenToTexture GETexture[0x04000000-0x04088000, 480x272 (texture 1024x1024), bufferWidth=512, pixelFormat=3(PSM_8888)]
04:09:34 DEBUG  runtime - user_main - Start of Callback 0x089D1F94
04:09:34 DEBUG hle.sceDisplay - GUI - sceDisplay.paintGL - rendering the FB 0x04088000
04:09:34 DEBUG hle.ThreadManForUser - user_main - sceKernelSignalSema semaid=0x6A, signal=0x1
04:09:34 DEBUG       ge - GUI - GETexture.copyTextureToScreen GETexture[0x04088000-0x04110000, 480x272 (texture 1024x1024), bufferWidth=512, pixelFormat=3(PSM_8888)] at 0x0
04:09:34 DEBUG hle.ThreadManForUser - user_main - onSemaphoreModified waking thread user_main(uid=0x10, Status=PSP_THREAD_WAITING, Priority=0x20, Wait=Semaphore (0x006A) (delay 5000000 us, rest 4992326 us), doCallbacks=false)
04:09:34 DEBUG hle.sceDisplay - GUI - sceDisplay.paintGL - end display
04:09:34 DEBUG hle.ThreadManForUser - user_main - Context switching to 'CRI Wave out(uid=0x78, Status=PSP_THREAD_READY, Priority=0x10, Wait=None, doCallbacks=false)' after reschedule
04:09:34 DEBUG hle.ThreadManForUser - user_main - Inside an interrupt, not context switching to CRI Wave out(uid=0x78, Status=PSP_THREAD_READY, Priority=0x10, Wait=None, doCallbacks=false)
04:09:34 DEBUG hle.ThreadManForUser - user_main - hleKernelSignalSema returning 0, SceKernelSemaInfo(uid=0x6A, name='igSyncFifoSequence', attr=0x0, currentCount=0, maxCount=1, numWaitThreads=0)
04:09:34 DEBUG hle.ThreadManForUser - user_main - sceKernelSignalSema returning 0x0
04:09:34 DEBUG  runtime - user_main - End of Callback 0x089D1F94
04:09:34 DEBUG hle.sceGe_user - user_main - hleGeOnAfterCallback restarting list PspGeList[id=0x13, status=PSP_GE_LIST_DONE, list=0x08F19980, pc=0x08F1999C, stall=0x00000000, cbid=0x0, ended=true, finished=true, paused=true, restarted=false, reset=false]
04:09:34 DEBUG hle.ThreadManForUser - user_main - End of Interrupt
04:09:34 DEBUG hle.ThreadManForUser - user_main - Context switching to 'CRI Wave out(uid=0x78, Status=PSP_THREAD_READY, Priority=0x10, Wait=None, doCallbacks=false)' after reschedule
04:09:34 DEBUG hle.ThreadManForUser - user_main - ----- CRI Wave out(uid=0x78, Status=PSP_THREAD_RUNNING, Priority=0x10, Wait=None, doCallbacks=false), now=50703866
04:09:34 DEBUG  runtime - user_main - Switching from Thread idle1 to CRI Wave out
04:09:34 DEBUG  runtime - user_main - Ending Idle State
04:09:34 DEBUG  runtime - user_main - syncThread currentThread=user_main, currentRuntimeThread=CRI Wave out
04:09:34 DEBUG  runtime - user_main - Waiting to be scheduled...
```

**Comment by gid15** [comment]
There is only 1 list processing in the log file and no call to sceGe and sceDisplay. I guess that when enabling the DEBUG log, the "CRI Wave out" is keeping the PSP cpu busy and no longer scheduling to another PSP thread. This is sometimes the case. The audio threads are often the most active ones on a real PSP and consuming most the cpu power.

Could you use the following LogSettings.xml, I've tried to disable most of the logging and just keep the interesting parts at the DEBUG level:

https://gist.github.com/gid15/ae43771e9e9255c39191f4daf93c6c8a

This is enabling the DEBUG level from the beginning, so just go as quickly as possible to the interesting place in the game and post the created log.txt.

Thank you!

**Comment by gid15** [comment]
I think I have found a synchronization issue.

Could you try commenting out the lines 488 to 500 in RuntimeContext, i.e. avoid sleeping at all while being idle.

There is a synchronization issue when all the PSP threads are being idle: RuntimeContext.syncIdle is being executed in that case and it tries to be nice with the host CPU and sleeps as much as possible. For that, it checks when the next scheduled action need to take place and sleeps for that amount of time. However, if a GE list (running in a different thread) is being completed (FINISH/SIGNAL + END) during that time and a GE callback needs to be executed, the call to the GE callback will be added to the scheduler for immediate execution. However, the idle thread sleeping in RuntimeContext.syncIdle doesn't see this change in the next scheduled action: there is now one action that need to be executed right now, but the sleeping thread does not see that change and will awake some time later. As a consequence, the GE callback will be executed with some delay. The delay might be small, but in the case you described above, this will sum up for each displayed character (as each has its own GE list).
When removing the "sleep" from RuntimeContext.syncIdle, the GE callback will be executed immediately with the drawback that the CPU usage of the host will always be high, even when the PSP is doing nothing.

**Comment by soywiz** [comment]
WOW. Commenting those lines totally fixes that issue :) At least that game.
BTW the other day I pushed this branch:
https://github.com/jpcsp/jpcsp/tree/lwjgl-paint-hack

That would reduce changing gl context overhead.

**Comment by soywiz** [comment]
Sorry for not replying yesterday. I was focused on some other stuff and didn't check the mail by then.

Instead of sleeping, what about waiting for a signal with a timeout? And check queueing stuff you can signal it.

**Comment by soywiz** [comment]
BTW Tekken 6 ingame, is still slow commenting those lines. So maybe there is another issue. I will try to determine the problem here.
